Rover/mode_dock.cpp

```diff
@@ -0,0 +1,234 @@
+    AP_GROUPINFO("_HDG_CORR_WT", 4, ModeDock, hdg_corr_weight, 0.75f),
+
+    // @Param: _STOP_DIST
+    // @DisplayName: Distance from docking target when we should stop
+    // @Description: The vehicle starts stopping when it is this distance way from docking target
+    // @Units: m
+    // @Range: 0 2
+    // @Increment: 0.01
+    // @User: Standard
+    AP_GROUPINFO("_STOP_DIST", 5, ModeDock, stopping_dist, 0.30f),
+
+    AP_GROUPEND
+};
+
+ModeDock::ModeDock(void) : Mode()
+{
+    AP_Param::setup_object_defaults(this, var_info);
+}
+
+#define AR_DOCK_ACCEL_MAX              20.0    // acceleration used when user has specified no acceleration limit
+
+// initialize dock mode
+bool ModeDock::_enter()
+{
+#if PRECISION_LANDING == ENABLED
+    // refuse to enter the mode if dock is not in sight
+    if (!rover.precland.enabled() || !rover.precland.target_acquired()) {
+        return false;
+    }
+
+    if (hdg_corr_enable && is_negative(desired_dir)) {
+        // DOCK_DIR is required for heading correction
+        gcs().send_text(MAV_SEVERITY_ERROR, "Dock: Set DOCK_DIR or disable heading correction");
+        return false;
+    }
+
+    // set speed limit to dock_speed param if available
+    // otherwise the vehicle uses wp_nav default speed limit
+    const float speed_max = is_positive(speed) ? speed : g2.wp_nav.get_default_speed();
+    float atc_accel_max = MIN(g2.attitude_control.get_accel_max(), g2.attitude_control.get_decel_max());
+    if (!is_positive(atc_accel_max)) {
+        // accel_max of zero means no limit so use maximum acceleration
+        atc_accel_max = AR_DOCK_ACCEL_MAX;
+    }
+    const float accel_max = is_positive(g2.wp_nav.get_default_accel()) ? MIN(g2.wp_nav.get_default_accel(), atc_accel_max) : atc_accel_max;
+    const float jerk_max = is_positive(g2.wp_nav.get_default_jerk()) ? g2.wp_nav.get_default_jerk() : accel_max;
+
+    // initialise position controller
+    g2.pos_control.set_limits(speed_max, accel_max, g2.attitude_control.get_turn_lat_accel_max(), jerk_max);
+    g2.pos_control.init();
+
+    // check if we need to dock in reverse direction
+    _reversed = rover.precland.get_orient() == 4;
+    g2.pos_control.set_reversed(_reversed);
+
+    _loiter_after_docking = rover.is_boat();
+    _docking_complete = false;
+
+    return true;
+#else
+    return false;
+#endif
+}
+
+void ModeDock::update()
+{
+#if PRECISION_LANDING == ENABLED
+    // if docking is complete, rovers stop and boats loiter
+    if (_docking_complete) {
+        // rovers stop, boats loiter 
+        // note that loiter update must be called after successfull initialisation on mode loiter
+        if (_loiter_after_docking) {
+            // mode loiter must be initialised before calling update_post_docking method
+            rover.mode_loiter.update();
+        } else {
+            stop_vehicle();
+        }
+        return;
+    }
+
+    bool dock_in_sight = rover.precland.get_target_position_cm(_dock_pos_rel_origin_cm);
```

Comment:
```suggestion
    bool real_dock_in_sight = rover.precland.get_target_position_cm(_dock_pos_rel_origin_cm);
```

Maybe even move this closer to where it's being used - makes it clearer that there's the programmed vs seen positions.

Reply:
Changed the variable name. Can't move it closer to the place where it is used since we have to do some calculations based on `_dock_pos_rel_origin_cm` before we use that boolean.